### ISSUE TITLE
the sample code in fresh 1.2 release blog will not work

### ISSUE THREAD

**Issue by tanquar** [denoland/fresh#1342]
The sample code below at https://deno.com/blog/fresh-1.2 will not work.

```
// routes/index.tsx
export function Page() {
  const cart = useSignal<string[]>([]);
  return (
    <div>
      <Header cart={cart} />
      <div>
        <h1>Lemon</h1>
        <p>A very fresh fruit.</p>
        <AddToCart cart={cart} id="lemon" />
      </div>
    </div>
  );
}

// islands/Header.tsx
export default function Header(props: { cart: Signal<string[]> }) {
  return (
    <header>
      <span>Fruit Store</span>
      <button>Open cart ({props.cart.value.length})</button>
    </header>
  );
}

// islands/AddToCart.tsx
export default AddToCart(props: { cart: Signal<string[]>, id: string }) {
  function add() {
    props.cart.value = [...props.cart.value, id];
  }
  return <button onClick={add}>Add to cart</button>
}
```

At least, 
```
export default AddToCart(props: { cart: Signal<string[]>, id: string }) {
```
should accompany `function`:

```
export default function AddToCart(props: { cart: Signal<string[]>, id: string }) {
```

After adding it, the code still not work in my environment.
When I press "Add to cart" button, the browser console says "Uncaught TypeError: props.cart.value is not iterable".
It looks like `const cart = useSignal<string[]>([])` is not correctly passed to the AddToCart island, which is initialized on the server, and then (expected to be) passed to the client.

I wrote the following code, which is successful.
```
// routes/cart.tsx
import Cart from "~/islands/Cart/cart.tsx";

export default function Page(props: PageProps) {
  return (
    <div>
      <Cart />
    </div>
  );
}

// islands/Cart/cart.tsx
import { useSignal } from "@preact/signals";
import AddToCart from "~/islands/Cart/add.tsx";
import Header from "~/islands/Cart/header.tsx";

export default function Cart() {
  const cart = useSignal<string[]>([]);

  return (
    <div>
      <Header cart={cart} />
      <div>
        <h1>Lemon</h1>
        <p>A very fresh fruit.</p>
        <AddToCart cart={cart} id="lemon" />
      </div>
    </div>
  );
}

// islands/Cart/header.tsx
import { Signal } from "@preact/signals";

export default function Header(props: { cart: Signal<string[]> }) {
  return (
    <header>
      <span>Fruit Store</span>
      <button>Open cart ({props.cart.value.length})</button>
    </header>
  );
}

// islands/Cart/add.tsx
import { Signal } from "@preact/signals";

export default function AddToCart(props: { cart: Signal<string[]>, id: string }) {
  function add() {
    props.cart.value = [...props.cart.value, props.id];
  }
  return (<button onClick={add}>Add to cart</button>)
}
```

The above code works as I expect. Pressing the "Add to cart" button refreshes the "Open cart (0)" to "(1)".
I assume that I miss something that needs to be done for the original code to run as expected, or, the original code is only intended within a large island (as in my sample code).

Question: What is the expected behavior of useSignal within routes, then passed to an island?

**Comment by sant123** [comment]
I'm having the same issue, creating a signal in a route makes the behavior different between `server` and `client` side rendering. This is my signal in index.tsx:

```tsx
export default function Page() {
  const isDark = useSignal(true);

  return (
    <>
      <form action="/connect" class="m-4 flex items-center">
        <DarkMode isDark={isDark} />
      </form>
    </>
  );
}
```

And my island DarkMode.tsx

```tsx
export default function DarkMode({ isDark }: Props) {
  console.log(isDark);
```

Server prints:

![image](https://github.com/denoland/fresh/assets/7959437/684f8793-2e43-4524-985e-16713b901c0f)

But client:

![image](https://github.com/denoland/fresh/assets/7959437/7a9acf52-627c-493b-ac49-4c6eb2fe0d95)

So `isDark` is no longer a signal in the client, so trying to set the "signal" will fail with:

![image](https://github.com/denoland/fresh/assets/7959437/86cec9e1-9e75-48e1-80dc-2a403a675884)

**Comment by marvinhagemeister** [comment]
@tanquar Thanks for sharing this issue. That's indeed an error in the code snippet in the release blog post. Made a PR to address this and it will probably land later today. Thanks again for bringing this up!

@sant123 I can't reproduce the behaviour you're describing. When I try out the provided code snippet the client logs the signal object instead of the unwrapped boolean value for me. Do the snippets contain the exact code you used to see that behaviour?

**Comment by marvinhagemeister** [comment]
The code snippet in the blog post is fixed 🎉

**Comment by tanquar** [comment]
Thank you!
I'm looking at the actual blog post at https://deno.com/blog/fresh-1.2 and I tested with a blank fresh 1.2 project, which was successful. Server side `const cart = useSignal<string[]>([])` in "routes" is passed to the client side islands.

The same cart sample code still didn't work in my existing project, upgraded from fresh 1.1 to 1.2.
I figured out the different version with signals-core in import_map:
```
@preact/signals-core": "https://esm.sh/*@preact/signals-core@1.2.3", // fresh 1.2
@preact/signals-core": "https://esm.sh/*@preact/signals-core@1.3.0", // my existing project
```

I used "1.3.0" which looks like the latest, and the current fresh `deno run -A -r https://fresh.deno.dev` uses "1.2.3". After I downgraded to "signals-core@1.2.3", the cart code worked in my existing project successfully.

**Comment by sant123** [comment]
> Thank you! I'm looking at the actual blog post at https://deno.com/blog/fresh-1.2 and I tested with a blank fresh 1.2 project, which was successful. Server side `const cart = useSignal<string[]>([])` in "routes" is passed to the client side islands.
> 
> The same cart sample code still didn't work in my existing project, upgraded from fresh 1.1 to 1.2. I figured out the different version with signals-core in import_map:
> 
> ```
> @preact/signals-core": "https://esm.sh/*@preact/signals-core@1.2.3", // fresh 1.2
> @preact/signals-core": "https://esm.sh/*@preact/signals-core@1.3.0", // my existing project
> ```
> 
> I used "1.3.0" which looks like the latest, and the current fresh `deno run -A -r https://fresh.deno.dev` uses "1.2.3". After I downgraded to "signals-core@1.2.3", the cart code worked in my existing project successfully.

I’m using the 1.3.0 as well 👀

**Comment by marvinhagemeister** [comment]
Uh ok that sounds suspicious. Re-opening the issue, because something seems to be going on with the different signal versions. Will look into that tomorrow

**Comment by marvinhagemeister** [comment]
I think I found the root cause. To serialize props we're relying on `JSON.stringify` with a custom replacer function. In `@preact/signals-core 1.3.0` we added a `toJSON` method for folks who want to serialize signals to json for data exchange purposes. This breaks our use case though because that returns the signal's value so our serializer is blind to it having been a signal. Reason for that is that `JSON.stringify` calls an objects `.toJSON` method if present.
